### ISSUE TITLE
App freeze when tap songs to play on IOS

### ISSUE THREAD

**Issue by BechirAhmed** [Sangwan5688/BlackHole#139]
**Describe the bug**

The last version of the app, whenever I click on song to play the app freeze, and I have to restart it, I've tried a lot.
I've set use equalizer to false, but still the same issue

**Smartphone (please complete the following information):**
 - Device: [iPhone Xs-Max]
 - OS: [e.g. IOS 15.1]
 - App Version [last version]

**App Logs**
If possible please provide app logs here.

**Comment by BechirAhmed** [comment]
I've cloned the project v1.8.0 and build it using my mac and it was working fine on IOS 15 & Android, but when I cloned the last version of the project v1.11.0 it open the app and show all songs and playlists, but when I click on song to play it just freeze on both IOS & Android, and it didn't give any logs

**Comment by Sangwan5688** [comment]
Ummm maybe coz of your flutter version or something. Make sure to use latest flutter version available. As of now it should be above flutter 2.5. Then make sure you are using the defined package version. It should work fine then.
Also, if you build iOS versions from the app. I do request to contribute the same to the project by creating Pull Requests. That way others can run the app on their iOS devices too. Coz I can't create iOS apk as of now.

**Comment by BechirAhmed** [comment]
Well i will recheck the versions again, thanks for your time and the huge work.
About IOS PRs, I will, as soon as I get it working well.

I'm still looking for alternatives packages for (ext_storage, equalizer, display_mode..) since those packages works only on android.

@Sangwan5688 One last question, why you didn't use only packages that works on both OSs? Even you don't have mac, At least you will make sure that everyone can build the project successfully

**Comment by Sangwan5688** [comment]
Ya I do make sure to use packages which are available on as many platforms as possible. But you can't always find good packages which are available on both. For ext_storage you will have to do a workaround. Use path_provider to get external storage path then edit that part to get desired path. For display_mode, you don't need to use that in iOS. So you don't need that package in iOS. You can also remove Equalizer for iOS.

**Comment by Sangwan5688** [comment]
If you have anything else regarding iOS please continue the same in #42

**Comment by BechirAhmed** [comment]
Hi @Sangwan5688 , About this issue, the app actually doesn’t freeze, but instead it return an empty SizedBox because the audioHanlder.mediaItem always null, I don't know how, in audioPlayer screen, do you have any idea?
```
StreamBuilder<MediaItem>(
          stream: audioHandler.mediaItem,
          builder: (context, snapshot) {
            final MediaItem mediaItem = snapshot.data;
            print(mediaItem); //Always null
            if (mediaItem == null) return SizedBox();

```

**Comment by samyakkkk** [comment]
@BechirAhmed were you able to fix the play button issue? we have been trying to fix it the whole day but couldn't actually understand the problem. It looks like that all the functions called on AudioPlayer don't revert any response back and the media item is never updated.

**Comment by BechirAhmed** [comment]
@samyakkkk the issue i hade is when hitting a song to play nothing happen, firstly I thought that the app freezes but actually id doesn’t , Later I figured out how to solve the issue, I'll try to describe the issue and @Sangwan5688 can tell us if this is the right way to solve it.

```
final MediaItem? mediaItem = snapshot.data;
if (mediaItem == null) return const SizedBox();
``` 
This first line always giving null, I don't know why, So obviously the following condition will always returning Empty SizedBox which will block the screen, and to confirm that just swipe down the screen from the top edge and you can manipulate the app again.

To solve that i stoped audioHandler in initState() and it works.
```
if (widget.offline == null) {
      if (audioHandler.mediaItem.value?.extras!['url'].startsWith('http')
          as bool) {
        offline = false;
      } else {
        offline = true;
      }
    } else {
      offline = widget.offline!;
    }
    audioHandler.stop(); /////////////////////////////////////// PUT THIS LINE
    fromMiniplayer = widget.fromMiniplayer;

```

**Comment by Sangwan5688** [comment]
Umm, Ya @BechirAhmed is right about that screen freezing issue. I thought I already told you about it but I guess I never did in this issue. Maybe I told it in the windows issue or maybe somewhere else. Anyway, whenever there's no mediadata returned which mostly happens when the package doesn't support any particular platform then a sizedbox is return. As the playscreen has a transparent background and just a sizedbox is return it appears to be frozen. But there's just a transparent layer.
As for the solution suggested, I think stopping audiohandler won't do the work. I don't know what you meant by `... and it works.` It starts playing or what? Most probably not, it will just remove that transparent player afaik. Also, this might work on platforms where it misbehaves but for platforms where it was already working? Will it not cause any issue there? Also, even it there's no issue the stopped audioHandler will then need to restart which will increase the loading time. So, I definitely won't prefer that.
Now, coming over to why actually it's showing empty screen. See `audio_service` does support iOS so it's definitely not the case of package not supporting anything. Most probably either that particular method/function/stream which is being listened is not supported on that platform or there's something wrong in the implementation for that particular platform. You can check about the package [audio_service](https://pub.dev/packages/audio_service) and [just_audio](https://pub.dev/packages/just_audio) and see of there's something missing in iOS side or not.

**Comment by BechirAhmed** [comment]
Yes @Sangwan5688 it starts playing, and yes stopping audioHandler may not be the right solution but it did it. I've tried to print that mediaItem just before the condition and without audioHandler.stop() it print **null** **null** twice. 
And after adding audioHandler.stop() it print **null** and then printing the mediaItem. As i said i didn't understand why but i think it related to an issue i have reported it last time about iOS entitlement, there's something unsupported in audio_service package

**Comment by samyakkkk** [comment]
@BechirAhmed thanks for helping us on this. adding the audioHandler.stop() did get this working for us in both iOS and Android. It might not be the best way but I don't see any side effects as of now.

**Comment by agbekodzamesi** [comment]
> @samyakkkk the issue i hade is when hitting a song to play nothing happen, firstly I thought that the app freezes but actually id doesn’t , Later I figured out how to solve the issue, I'll try to describe the issue and @Sangwan5688 can tell us if this is the right way to solve it.
> 
> ```
> final MediaItem? mediaItem = snapshot.data;
> if (mediaItem == null) return const SizedBox();
> ```
> 
> This first line always giving null, I don't know why, So obviously the following condition will always returning Empty SizedBox which will block the screen, and to confirm that just swipe down the screen from the top edge and you can manipulate the app again.
> 
> To solve that i stoped audioHandler in initState() and it works.
> 
> ```
> if (widget.offline == null) {
>       if (audioHandler.mediaItem.value?.extras!['url'].startsWith('http')
>           as bool) {
>         offline = false;
>       } else {
>         offline = true;
>       }
>     } else {
>       offline = widget.offline!;
>     }
>     audioHandler.stop(); /////////////////////////////////////// PUT THIS LINE
>     fromMiniplayer = widget.fromMiniplayer;
> ```

Have implemented this fix and it works.

**Comment by Sangwan5688** [comment]
I see if it's working fine then you can implement it. But just do one thing. Make sure to check the platform and if platform is android then don't call audioHandler.stop() whereas for rest platforms you can call .stop() in init function. And If everything is working fine in iOS side now, can you build its app and upload in #42. I don't actually know how to sideload apps on iOS but I hope you can sideload them.

**Comment by BechirAhmed** [comment]
@Sangwan5688 Is there any side effects in Android if called audioHandler.stop()?

**Comment by Sangwan5688** [comment]
See audioHandler.stop() is called when you no longer want to play audio. It doesn't release the codecs but it removes audio service from foreground. It is called whenever you swipe away the miniplayer or tap that cross icon from notification and app is open. Usually when play screen is initialized you don't want previous queue so it won't cause any issue with queue but there can be some cases where it might cause some issue. First of all opening app from miniplayer, when you open app from miniplayer play screen is initialized so it will stop the song in that case, causing issues. Except that I don't think there can be any other issue. So a check for checking if it's from miniplayer or not is required in those cases. Now coming over to performance, see whenever you call audoHandler.stop() then resources are released. Normally app just updates the current playing queue but calling audioHandler.stop() will cause the service to release resources and remove it from foreground service. So instead of remaining in foreground and changing queue, app will keep removing from foreground service too. It won't release the codecs so process won't be that slow but still it will be slower as unnecessary steps are done.
So in the end i would suggest not to do it for android. As it solves the current problem with iOS so you can do so for iOS but keep android out of it. Coming over to other platforms like windows, you can do it for windows or other desktop systems too as in desktops there's no foreground service so it won't cause any issue. Just make sure to add that miniplayer check.

So, in the the suggested changes should look like this:
```
...
....
if (!platform.isAndroid && !fromMiniplayer){
    // you can check it with miniplayer too
    // if there's no issue even when it is from miniplayer
    // then you can remove fromMiniplayer part too
    audiohandler.stop();
}
```

**Comment by Sangwan5688** [comment]
So again thanks to @BechirAhmed for pointing out that playback works fine on stopping the audioHandler before playing any song. Again, I don't know why it works but well as long as it works its fine.
I have implemented the same in latest commit to make sure nothing is changed in Android side and also made sure that opening from miniplayer doesn't cause the music to stop. If anything doesn't work as expected then do correct me and push changes to the repo
